db_stress_tool/expected_state.cc

```diff
@@ -3,39 +3,106 @@
+  const bool during_read_no_delete_happened =
+      (pre_read_expected_del_counter ==
+       GetDelCounter(post_read_expected_del_counter +
+                     expected_del_counter_delta));
+
+  return pre_read_expected_not_deleted && during_read_no_delete_happened;
+}
+
+bool ExpectedState::InExpectedValueBaseRange(
+    uint32_t value_base, uint32_t pre_read_expected_value_base,
+    uint32_t post_read_expected_value_base,
+    bool post_read_expected_pending_write) {
+  assert(IsValuePartValid(value_base, VALUE_BASE_MASK) &&
+         IsValuePartValid(pre_read_expected_value_base, VALUE_BASE_MASK) &&
+         IsValuePartValid(post_read_expected_value_base, VALUE_BASE_MASK));
+
+  const uint32_t expected_value_base_delta =
+      post_read_expected_pending_write ? VALUE_BASE_DELTA : 0;
+
+  uint32_t post_read_expected_value_base_final =
+      GetValueBase(post_read_expected_value_base + expected_value_base_delta);
+
+  if (pre_read_expected_value_base <= post_read_expected_value_base_final) {
+    const uint32_t lower_bound = pre_read_expected_value_base;
+    const uint32_t upper_bound = post_read_expected_value_base_final;
+    return lower_bound <= value_base && value_base <= upper_bound;
+  } else {
+    const uint32_t upper_bound_1 = post_read_expected_value_base_final;
+    const uint32_t lower_bound_2 = pre_read_expected_value_base;
+    const uint32_t upper_bound_2 = VALUE_BASE_MASK;
+    return (value_base <= upper_bound_1) ||
+           (lower_bound_2 <= value_base && value_base <= upper_bound_2);
+  }
+}
+
 ExpectedState::ExpectedState(size_t max_key, size_t num_column_families)
     : max_key_(max_key),
       num_column_families_(num_column_families),
       values_(nullptr) {}
 
 void ExpectedState::ClearColumnFamily(int cf) {
   std::fill(&Value(cf, 0 /* key */), &Value(cf + 1, 0 /* key */),
-            SharedState::DELETION_SENTINEL);
+            ExpectedState::DEL_MASK);
 }
 
-void ExpectedState::Put(int cf, int64_t key, uint32_t value_base,
-                        bool pending) {
-  if (!pending) {
-    // prevent expected-value update from reordering before Write
-    std::atomic_thread_fence(std::memory_order_release);
+void ExpectedState::Put(int cf, int64_t key, uint32_t value, bool pending) {
+  uint32_t new_value = Value(cf, key).load();
+  if (PendingDelete(new_value)) {
+    // Needed when a crash has happened in the middle of the last deletion
+    ClearValuePart(&new_value, PENDING_DEL_MASK);
   }
-  Value(cf, key).store(pending ? SharedState::UNKNOWN_SENTINEL : value_base,
-                       std::memory_order_relaxed);
   if (pending) {
-    // prevent Write from reordering before expected-value update
-    std::atomic_thread_fence(std::memory_order_release);
+    SetValuePart(&new_value, PENDING_WRITE_MASK, PENDING_WRITE_MASK);
+    Value(cf, key).store(new_value);
+  } else {
+    SetValuePart(&new_value, VALUE_BASE_MASK, GetValueBase(value));
```

Comment:
Maybe these should be methods on an `ExpectedValue` class that owns a uint32_t

Reply:
> It seems clearer if value were not provided at all, and here you use NextValueBase(GetValueBase(new_value))


Are you suggesting we have something like below in `Status TestPut()`

```
    const uint32_t value_base =
        ExpectedValueHelper::NextValueBase(ExpectedValueHelper::GetValueBase(
            shared->Get(rand_column_family, rand_key)));
    const size_t sz = GenerateValue(value_base, value, sizeof(value));
    const Slice v(value, sz);

    shared->Put(rand_column_family, rand_key, true /* pending */); // Right now it's shared->Put(rand_column_family, rand_key value_base, true /* pending */); 
```

I decided to still receive `value_base` in `ExpectedState::Put()` mainly for better semantics with `const size_t sz = GenerateValue(value_base, value, sizeof(value));`, showing explicitly we use same `value_base` across `GenerateValue()` and  `shared->Put()`

I don't have a strong opinion for either way so just want to ensure I understand your suggestion.

Reply:
> It seems clearer if `value` were not provided at all, and here you use `NextValueBase(GetValueBase(new_value))`

I meant like the way it's done in `ExpectedState::Delete()`: https://github.com/facebook/rocksdb/pull/11058/files#diff-274c08b6c1955e222d9ffb3ebe1ecfbf2e94a0e8bc2417ff2d0b05ba9378d264R133-R135

Reply:
BTW, I know this does not solve your problem of how to generate the value for the actual `Put()`, and am still thinking about how to solve that and make things more nicely layered. Right now `ExpectedState` sometimes has its changed decided externally and sometimes decides them internally.

Reply:
Got you - let me think about it

Reply:
> Right now ExpectedState sometimes has its changes decided externally and sometimes decides them internally.

I think renaming some function names is good enough: 
```
  const uint32_t value_base = shared->GetNextUnusedValueBase(rand_column_family, rand_key);
  const size_t sz = GenerateValue(value_base, value, sizeof(value));
  const Slice v(value, sz);
  shared->Put(rand_column_family, rand_key, true /* pending */);  // ExpectedState updated to be same as GetNextUnusedValueBase()
```


The only complicated scenario is in recovery cleanup, which is still fine as the `value_base_from_db` should be the GetNextUnusedValueBase() upon crash

```
bool VerifyOrSyncValue(int cf, int64_t key, const ReadOptions& /*opts*/,
                         SharedState* shared, const std::string& value_from_db,
                         std::string msg_prefix, const Status& s) const {
    if (shared->HasVerificationFailedYet()) {
      return false;
    }
    const uint32_t expected_value = shared->Get(cf, key);

    if (ExpectedValueHelper::PendingWrite(expected_value) ||
        ExpectedValueHelper::PendingDelete(expected_value)) {
      if (s.ok()) {
        // Value exists in db, update state to reflect that
        const Slice slice(value_from_db);
        const uint32_t value_base_from_db = GetValueBase(slice);
        shared->Put(cf, key, value_base_from_db, false /* pending */);
      } else if (s.IsNotFound()) {
        // Value doesn't exist in db, update state to reflect that
        shared->SingleDelete(cf, key, false /* pending */);
      }
      return true;
    }
```

Reply:
Trying out something in the latest commit https://github.com/facebook/rocksdb/pull/11058/commits/66df71deaa7f79ba88236175a9d482cb27fce71e

Conceptually it looks fine to me but will be running stress test on it to see if I miss any edge case (e.g, txn recovery cleanup)

Reply:
IMO there is a missed opportunity for encapsulation still. For example, you could make `ExpectedState` functions return `ExpectedValue` class instead of uint32_t, which could have methods like `IncValueBase(bool pending)`, `GetValueBase()`, `GetFinalValueBase()` (this one would apply pending update when there is one), etc.

Reply:
> ```
>   const uint32_t value_base = shared->GetNextUnusedValueBase(rand_column_family, rand_key);
>   const size_t sz = GenerateValue(value_base, value, sizeof(value));
>   const Slice v(value, sz);
>   shared->Put(rand_column_family, rand_key, true /* pending */);  // ExpectedState updated to be same as GetNextUnusedValueBase()
> ```

This introduces a coupling between GetNextUnusedValueBase() and Put(), which isn't ideal.

By the suggestion in my last comment it would look like:

```
ExpectedValue ev = shared->Put(rand_column_family, rand_key, true /* pending */);
const size_t sz = GenerateValue(ev.GetFinalValueBase(), value, sizeof(value));
const Slice v(value, sz);
```

Reply:
Fixed